Feature/amazon chronos interface

RigvedManoj opened the pull request:
Implements Chronos Forecaster from https://github.com/sktime/sktime/issues/6119

What does this implement/fix? Explain your changes.
This impementation adds a new forecaster (Amazon's Chronos).

Does your contribution introduce a new dependency? If yes, which one?
Amazon Chronos (https://github.com/amazon-science/chronos-forecasting)

What should a reviewer concentrate their feedback on?
Class names and API layout
Design
CheckEstimator 

Did you add any tests for the change?
Not yet

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, for added estimators: I've added myself and possibly to the `maintaners` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

yarnabrina commented on sktime/forecasting/chronos.py:
```diff
@@ -0,0 +1,161 @@
+
+    # tag values are "safe defaults" which can usually be left as-is
+    _tags = {
+        "python_dependencies": ["torch"],
```
I think chronos specific library  eeds to be added. Project readme suggests to install from Github only, but I'm not sure if sktime tag handler supports that or not.

@fkiraly @benHeid if we specify chronos directly, and add it as a .git dependency in pyproject.toml, will it work?

yarnabrina commented on sktime/forecasting/chronos.py:
```diff
@@ -0,0 +1,161 @@
+        "maintainers": ["RigvedManoj"],
+    }
+
+    def __init__(self, modelName, num_samples=20, temperature=1.0, top_k=50, top_p=1.0, args_list=None, kwargs_dict=None):
```
Please use snake_case for `modelName` too.

RigvedManoj replied:
changed

yarnabrina commented on sktime/forecasting/chronos.py:
```diff
@@ -0,0 +1,161 @@
+        "maintainers": ["RigvedManoj"],
+    }
+
+    def __init__(self, modelName, num_samples=20, temperature=1.0, top_k=50, top_p=1.0, args_list=None, kwargs_dict=None):
```
How do I interpret temperature? In a text generation model, I can understand where sometimes we may need creativity. In this context, what will 1 (max) and 0 (min) imply?

yarnabrina replied:
I think I'll have similar doubts for top-p and top-k as well.

yarnabrina commented on sktime/forecasting/chronos.py:
```diff
@@ -0,0 +1,161 @@
+        self : reference to self
+        """
+        import torch
+        from chronos import ChronosPipeline
```
My guess is as of now this is going to fail in CI, as there's no installation of this library. You need to add it in pyproject.toml and in dependency tag above. 

Since changes in pyproject.toml triggers a lot of jobs, I'll request you to consider commenting unnecessary jobs in CI file if possible. We'll revert this commenting part later of course.

yarnabrina commented on sktime/forecasting/chronos.py:
```diff
@@ -0,0 +1,161 @@
+            top_k=self.top_k,
+            top_p=self.top_p,
+        )
+        values = np.median(forecast[0].numpy(), axis=0)
```
This is mainly a compatibility question to @fkiraly : do we intend to use predict of all forecasters as median or mean? Depending on that, this line may needs to be changed.

yarnabrina commented on sktime/forecasting/chronos.py:
```diff
@@ -0,0 +1,161 @@
+        row_idx = fh.to_absolute_index(self.cutoff)
+        col_idx = self._y.index
+        '''
+        y_pred = pd.DataFrame(values)
```
Is this in sktime expected format? How are you planning to match horizon in indices?

yarnabrina commented on sktime/forecasting/chronos.py:
```diff
@@ -0,0 +1,161 @@
+    _tags = {
+        "python_dependencies": ["torch"],
+        "y_inner_mtype": "pd.Series",
+        "X_inner_mtype": "pd.DataFrame",
```
This can be removed I think, if we are not going to use X.

yarnabrina commented on sktime/forecasting/chronos.py:
```diff
@@ -0,0 +1,161 @@
+        "requires-fh-in-fit": False,
+        "authors": ["RigvedManoj"],
+        "maintainers": ["RigvedManoj"],
+    }
```
Does chronos support in sample predictions? Does it support missing values? Those can be useful tags here.

yarnabrina commented:
@RigvedManoj did you see https://github.com/sktime/sktime/pull/6205/checks?sha=ef9e609d3ef85f2fabf29fef3b61414b08a78575?

As of now, even if I approve CI it's immediately going to fail first with code quality issues, and then because chronos library will be missing. Please address these issues, other than aforementioned review comments.

RigvedManoj commented:
> @RigvedManoj did you see https://github.com/sktime/sktime/pull/6205/checks?sha=ef9e609d3ef85f2fabf29fef3b61414b08a78575?
> 
> As of now, even if I approve CI it's immediately going to fail first with code quality issues, and then because chronos library will be missing. Please address these issues, other than aforementioned review comments.

Sorry, I am trying to check whether the check estimator will run succesfully here. I am not able to run it in my local.

I thought changing draft pr to actual pr might trigger a build, but I don't see any run? is there something else I should do to trigger the check estimator for this PR?



I will also look at the reviews and resolve them.

yarnabrina commented:
> I am trying to check whether the check estimator will run succesfully here

CI/CD does not run `check_estimator` directly for all/modified estimators. It instead runs selective tests separately as part of each class.

> I thought changing draft pr to actual pr might trigger a build, but I don't see any run?

CI does not depend on PR status. For first time contributors, a member needs to approve a run. After one successful PR, it's no longer required.

> is there something else I should do to trigger the check estimator for this PR?

Yes, please see the review comments. I am not approving the CI because I know it's immediately going to fail first for linting reasons (see the above [link](https://github.com/sktime/sktime/pull/6205/checks?sha=ef9e609d3ef85f2fabf29fef3b61414b08a78575) I shared), and then for missing libraries. You need to resolve these first.

1. For linting, follow [this](https://www.sktime.net/en/latest/developer_guide/coding_standards.html#using-pre-commit).
2. For missing packages, you need to modify `pyproject.toml` (in `dl` extra) and update tags (in `python_dependencies` tag) in the new class. Sine your work is closely related to @benHeid's work on Hugging Face adapter, he may be able to give some more specific suggestions.

RigvedManoj commented:
> 2\. For missing packages, you need to modify `pyproject.toml` (in `dl` extra) and update tags (in `python_dependencies` tag) in the new class. Sine your work is closely related to @benHeid's work on Hugging Face adapter, he may be able to give some more specific suggestions.

Shouldn't this be added to "forecasting" and "all_extras" onstead of dl?

yarnabrina commented:
> > 2. For missing packages, you need to modify `pyproject.toml` (in `dl` extra) and update tags (in `python_dependencies` tag) in the new class. Sine your work is closely related to @benHeid's work on Hugging Face adapter, he may be able to give some more specific suggestions.
> 
> Shouldn't this be added to "forecasting" and "all_extras" instead of dl?

No. If you'll check, you will see torch is present only in dl dependency, and not in others for the specific reason that it is quite big and has lots of dependencies. We don't want to force all forecasting users of sktime to install all these if they don't need it.

RigvedManoj commented:
> > > 2. For missing packages, you need to modify `pyproject.toml` (in `dl` extra) and update tags (in `python_dependencies` tag) in the new class. Sine your work is closely related to @benHeid's work on Hugging Face adapter, he may be able to give some more specific suggestions.
> > 
> > 
> > Shouldn't this be added to "forecasting" and "all_extras" instead of dl?
> 
> No. If you'll check, you will see torch is present only in dl dependency, and not in others for the specific reason that it is quite big and has lots of dependencies. We don't want to force all forecasting users of sktime to install all these if they don't need it.

Sorry, I am a bit confused. torch is added to "dl" but chronos is a forecaster and so needs to be added to "forecasting" correct?

```python
forecasting = [
  "arch>=5.6,<6.4",
  "chronos @ git+https://github.com/amazon-science/chronos-forecasting.git",
  ```
  
  like this?

yarnabrina commented:
> > > > 2. For missing packages, you need to modify `pyproject.toml` (in `dl` extra) and update tags (in `python_dependencies` tag) in the new class. Sine your work is closely related to @benHeid's work on Hugging Face adapter, he may be able to give some more specific suggestions.
> > > 
> > > 
> > > Shouldn't this be added to "forecasting" and "all_extras" instead of dl?
> > 
> > 
> > No. If you'll check, you will see torch is present only in dl dependency, and not in others for the specific reason that it is quite big and has lots of dependencies. We don't want to force all forecasting users of sktime to install all these if they don't need it.
> 
> Sorry, I am a bit confused. torch is added to "dl" but chronos is a forecaster and so needs to be added to "forecasting" correct?
> 
> ```python
> forecasting = [
>   "arch>=5.6,<6.4",
>   "chronos @ git+https://github.com/amazon-science/chronos-forecasting.git",
> ```
> 
> like this?

No. This change is implying even if someone wants to use just ARIMA, installing forecasting extra is going to cause installation of torch, accelerate, transformers and all of their dependencies. This is huge, and we do not want that.

Yes, chronos is a forecaster, but it's also a forecaster that's deep learning based. To reduce the dependencies of most users, we thus isolate DL specific dependencies separately in dl extra. That's the same logic for which neuralforecast is in dl, for example.

RigvedManoj commented:
I am unsure of what the failed tests are. I have made some of the other changes in the reviews but I don't want to push till I resolve the tests? 

what are these failed tests?

benHeid commented:
> I am unsure of what the failed tests are. I have made some of the other changes in the reviews but I don't want to push till I resolve the tests?
> 
> what are these failed tests?

I identified two problems related to the doc test.
1. You should probably add # doctest: +SKIP after the import statement of `torch` in the doctoring. Since it seems that the doctests are always executed even if soft-dependencies are not installed.
2. In line 48 three dots are missing in the doctest.

RigvedManoj commented:
> > I am unsure of what the failed tests are. I have made some of the other changes in the reviews but I don't want to push till I resolve the tests?
> > what are these failed tests?
> 
> I identified two problems related to the doc test.
> 
> 1. You should probably add # doctest: +SKIP after the import statement of `torch` in the doctoring. Since it seems that the doctests are always executed even if soft-dependencies are not installed.
> 2. In line 48 three dots are missing in the doctest.

Got it, thanks!

I also see some assertion erros where arrays are unequal. I will check on those as well.

RigvedManoj commented:
> > > I am unsure of what the failed tests are. I have made some of the other changes in the reviews but I don't want to push till I resolve the tests?
> > > what are these failed tests?
> > 
> > 
> > I identified two problems related to the doc test.
> > 
> > 1. You should probably add # doctest: +SKIP after the import statement of `torch` in the doctoring. Since it seems that the doctests are always executed even if soft-dependencies are not installed.
> > 2. In line 48 three dots are missing in the doctest.
> 
> Got it, thanks!
> 
> I also see some assertion erros where arrays are unequal. I will check on those as well.

I have added DocTest and rectified the dots issue.

I also added the horizon indices for the predicted values y_pred which i believe should rectify the other failures.

RigvedManoj commented:
There were some failures because y_pred name was not set correctly. Resolved it.

RigvedManoj commented:
```python
def test_score(self, estimator_instance, n_columns, fh_int_oos):
        """Check score method."""
        y = _make_series(n_columns=n_columns)
        y_train, y_test = temporal_train_test_split(y)
        estimator_instance.fit(y_train, fh=fh_int_oos)
        y_pred = estimator_instance.predict()
    
        fh_idx = check_fh(fh_int_oos).to_indexer()  # get zero based index
        expected = mean_absolute_percentage_error(
            y_test.iloc[fh_idx], y_pred, symmetric=False
        )
    
        # compare expected score with actual score
        actual = estimator_instance.score(y_test.iloc[fh_idx], fh=fh_int_oos)
>       assert actual == expected
E       assert 0.12403076830396009 == 0.17163778419301146
```


I am not able to figure out what is being tested here?

benHeid commented:
> ```python
> def test_score(self, estimator_instance, n_columns, fh_int_oos):
>         """Check score method."""
>         y = _make_series(n_columns=n_columns)
>         y_train, y_test = temporal_train_test_split(y)
>         estimator_instance.fit(y_train, fh=fh_int_oos)
>         y_pred = estimator_instance.predict()
>     
>         fh_idx = check_fh(fh_int_oos).to_indexer()  # get zero based index
>         expected = mean_absolute_percentage_error(
>             y_test.iloc[fh_idx], y_pred, symmetric=False
>         )
>     
>         # compare expected score with actual score
>         actual = estimator_instance.score(y_test.iloc[fh_idx], fh=fh_int_oos)
> >       assert actual == expected
> E       assert 0.12403076830396009 == 0.17163778419301146
> ```
> 
> I am not able to figure out what is being tested here?

 It is testing if the correctness of the score method by comparing the score with a mape calculated on a forecast of the method. Since the forecaster is probably not deterministic, this fails. To fix it,  I would introduce a parameter `deterministic` which set the random_state to a fixed value if True. Additionally, it is then required to set this parameters in all test_params sets. 
 
 For setting the random state. Probably you can use the set_seed function of transformers. You could take a look into the hugging face Adapter. There I did the same for the Informer/Autoformer/TimeSeriesModelTransformer.

Z-Fran commented:
I am planning to continue to work on this PR.

fkiraly commented:
what are your planned next steps?

Z-Fran commented:
> what are your planned next steps?

I have reproduced the error of CI locally. I'm trying to find the reason.

pranavvp16 commented:
I'm planning to continue working on this.

fkiraly commented:
@Z-Fran, are you stlil working on this? If yes, kindly update with findings or make PR with your most recent code.

Z-Fran commented:
> @Z-Fran, are you stlil working on this? If yes, kindly update with findings or make PR with your most recent code.

I commit a draft version of chronos here. https://github.com/sktime/sktime/pull/6665

fkiraly commented:
superseded by https://github.com/sktime/sktime/pull/7001
